src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -310,37 +309,124 @@ internal static class ConstraintsHelper
+                binder = binder.WithAdditionalFlags(BinderFlags.GenericConstraintsClause | BinderFlags.SuppressConstraintChecks | BinderFlags.SuppressTypeArgumentBinding);
 
-                return defaultClauses.ContainsOnlyEmptyConstraintClauses() ? ImmutableArray<TypeParameterConstraintClause>.Empty : defaultClauses;
+                var discarded = DiagnosticBag.GetInstance(); // We will recompute this diagnostics more accurately later, when binding without BinderFlags.SuppressTypeArgumentBinding  
+                clauses = binder.BindTypeParameterConstraintClauses(containingSymbol, typeParameters, typeParameterList, constraintClauses,
+                                                                    discarded, performOnlyCycleSafeValidation: true);
+                discarded.Free();
+
+                clauses = AdjustConstraintKindsBasedOnConstraintTypes(containingSymbol, typeParameters, clauses);
             }
 
-            // Wrap binder from factory in a generic constraints specific binder
-            // to avoid checking constraints when binding type names.
-            Debug.Assert(!binder.Flags.Includes(BinderFlags.GenericConstraintsClause));
-            binder = binder.WithAdditionalFlags(BinderFlags.GenericConstraintsClause | BinderFlags.SuppressConstraintChecks);
+            if (clauses.All(clause => clause.Constraints == TypeParameterConstraintKind.None))
+            {
+                clauses = ImmutableArray<TypeParameterConstraintClause>.Empty;
+            }
+
+            return clauses.SelectAsArray(clause => clause.Constraints);
+        }
+
+        internal static ImmutableArray<TypeParameterConstraintClause> AdjustConstraintKindsBasedOnConstraintTypes(Symbol container, ImmutableArray<TypeParameterSymbol> typeParameters, ImmutableArray<TypeParameterConstraintClause> constraintClauses)
+        {
+            int arity = typeParameters.Length;
+
+            Debug.Assert(constraintClauses.Length == arity);
+
+            Dictionary<TypeParameterSymbol, bool> isValueTypeMap = TypeParameterConstraintClause.BuildIsValueTypeMap(container, typeParameters, constraintClauses);
+            Dictionary<TypeParameterSymbol, bool> isReferenceTypeFromConstraintTypesMap = TypeParameterConstraintClause.BuildIsReferenceTypeFromConstraintTypesMap(container, typeParameters, constraintClauses);
+            ArrayBuilder<TypeParameterConstraintClause> builder = null;
+
+            for (int i = 0; i < arity; i++)
+            {
+                var constraint = constraintClauses[i];
+                var typeParameter = typeParameters[i];
+                TypeParameterConstraintKind constraintKind = constraint.Constraints;
+
+                Debug.Assert((constraintKind & (TypeParameterConstraintKind.ValueTypeFromConstraintTypes | TypeParameterConstraintKind.ReferenceTypeFromConstraintTypes)) == 0);
+
+                if ((constraintKind & TypeParameterConstraintKind.AllValueTypeKinds) == 0 && isValueTypeMap[typeParameter])
```

Comment:
It may be a bit simpler and more regular to move the `AllValueTypeKinds` check out of `BuildIsValueTypeMap` (and rename it `BuildIsValueTypeFromConstraintTypesMap`). Then we can just do `if (isValueTypeMap[typeParameter]) ...` here, which is similar to reference-type case below.

Reply:
> It may be a bit simpler and more regular to move the AllValueTypeKinds check out of BuildIsValueTypeMap (and rename it BuildIsValueTypeFromConstraintTypesMap). Then we can just do if (isValueTypeMap[typeParameter]) ... here, which is similar to reference-type case below.

We don't need to calculate anything based on types if any flag from AllValueTypeKinds is set. I see no good reason for performing an unnecessary work when it can easily be avoided.


---
In reply to: [520161050](https://github.com/dotnet/roslyn/pull/49227#discussion_r520161050) [](ancestors = 520161050)

Reply:
Makes sense. Does the same rationale apply to reference types? 
(ie. can we  change `BuildIsReferenceTypeFromConstraintTypesMap` to be "BuildIsReferenceTypeMap" and have it avoid calculating anything based on types when `AllReferenceTypeKinds` is set?)

---
In reply to: [520163425](https://github.com/dotnet/roslyn/pull/49227#discussion_r520163425) [](ancestors = 520163425,520161050)

Reply:
> Does the same rationale apply to reference types

No, it doesn't. They are not similar this way.

---
In reply to: [520180924](https://github.com/dotnet/roslyn/pull/49227#discussion_r520180924) [](ancestors = 520180924,520163425,520161050)